adaptive_scheduler/kernel_mappings.py

```diff
@@ -170,25 +135,22 @@ def get_rise_set_timepoint_intervals(rise_set_target, visibility, max_airmass, m
     return dark_intervals
 
 
-def construct_compound_reservation(request_group, semester_start):
+def construct_compound_reservation(request_group, semester_start, network_model):
     '''Convert a RequestGroup into a CompoundReservation, translating datetimes
        to kernel epoch times. The Request windows were already translated into visible windows during the 
        filter_on_visibility step.
     '''
     reservations = []
     for index, request in enumerate(request_group.requests):
-        visibility_intervals_for_resources = req_windows_to_kernel_intervals(request.windows.windows_for_resource)
-        kernel_intervals_for_resources = translate_request_windows_to_kernel_windows(visibility_intervals_for_resources,
-                                                                                     semester_start)
+        kernel_intervals_for_resources = request.windows.to_kernel_intervals(semester_start)
+
+        # If the request has an optimization type of AIRMASS, pre-calculate and cache the airmasses at epoch values here. 
+        if request.optimization_type == 'AIRMASS':
+            request.cache_airmasses_within_kernel_windows(kernel_intervals_for_resources, network_model, semester_start)
```

Comment:
This is the point where we pre-compute and cache the airmasses for each request. This happens each run before we get to the scheduling kernel, but after we have already applied all pre-filters to the visibility windows. This was the best place I could find to do the caching since most of the information we need is available.